### ISSUE TITLE
Memory leak with PyTorch

### ISSUE THREAD

**Issue by gvtulder** [h5py/h5py#2010]
* Operating System: **Ubuntu 20.04.3 LTS**
 * Python version: **Python 3.8.10**
 * Where Python was acquired: **system**
 * h5py version: **3.4.0** to **3.6.0** (but only when installed with `pip`, not when compiled from git)
 * PyTorch version: **1.10.0+cu102** (installed with `pip install torch`)
 * Summary of the h5py configuration:
```---------------------------------
h5py    3.6.0
HDF5    1.12.1
Python  3.8.10 (default, Sep 28 2021, 16:10:42) 
[GCC 9.3.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.21.4
cython (built with) 0.29.24
numpy (built against) 1.17.3
HDF5 (built against) 1.12.1
```

Summary
=======
The Python memory explodes when an array is loaded from an h5py object and converted to a PyTorch tensor, but only if you also retrieve an attribute. This was not a problem in h5py <= 3.3.0, but it is a problem for 3.4.0, 3.5.0, and 3.6.0.

Steps to reproduce
===============
Install the packages with `pip install h5py numpy torch` and run:
```python
import h5py
import numpy as np
import torch

with h5py.File('test.h5', 'w') as ds:
    ds.create_dataset('scan', data=np.ones(shape=(1000, 1000)))
    ds['scan'].attrs['example'] = 'abcdefg'

def load_leaky():
    with h5py.File('test.h5', 'r') as ds:
        x_i = ds['scan'][:]
        ds['scan'].attrs['example']
    torch.tensor(x_i)

while True:
    load_leaky()
```

Workaround
=========
It works fine if you make a copy of the array before giving it to torch:
```python
        x_i = ds['scan'][:].copy()
```

Additional remarks
==============
* The `attrs` line is necessary. If you don't retrieve the attribute it works fine.
* The `torch.tensor` conversion is necessary to create a memory leak, but since older h5py versions work fine with the same version of PyTorch, and since it depends on loading the attribute, it is perhaps not a problem in PyTorch.
* For some reason, this is only a problem if I install the precompiled h5py packages with `pip`. I cloned the repository and used `dev-install.sh` to compile my own version, but then the problem did not appear.

**Comment by tacaswell** [comment]
What version of libhdf5 does it leak / not leak with?  Does throwing a `gc.collect()` in the loop help?

Given that you are opening the file every time, all of the libhdf5 caches should be torn down.

**Comment by gvtulder** [comment]
`gc.collect()` in the loop does not help.

The problem seems to appear between libhdf5 1.12.0 and 1.12.1:
* The pip package for h5py 3.3.0 (not leaking) was compiled with libhdf5 1.12.0, and libhdf5 1.12.1 is used from h5py 3.4.0 onwards (leaking).
* I compiled h5py 3.6.0 from git, using either libhdf5 1.12.0 (no leaking) or libhdf5 1.12.1 (leaking).

The libhdf5 that is installed on the system is 1.10.4+repack-11ubuntu1, the Ubuntu default. I am not sure whether h5py is using that or the libhdf5 that it was compiled with.

Compiled with libhdf5 1.21.1 (leaking):
```
python -c 'import h5py; print(h5py.version.info)'
Summary of the h5py configuration
---------------------------------
h5py    3.6.0
HDF5    1.12.1
Python  3.8.10 (default, Sep 28 2021, 16:10:42) 
[GCC 9.3.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.21.4
cython (built with) 0.29.24
numpy (built against) 1.21.4
HDF5 (built against) 1.12.1
```

Compiled with libhdf5 1.21.0 (not leaking):
```
python -c 'import h5py; print(h5py.version.info)'
Summary of the h5py configuration
---------------------------------
h5py    3.6.0
HDF5    1.12.0
Python  3.8.10 (default, Sep 28 2021, 16:10:42) 
[GCC 9.3.0]
sys.platform    linux
sys.maxsize     9223372036854775807
numpy   1.21.4
cython (built with) 0.29.24
numpy (built against) 1.21.4
HDF5 (built against) 1.12.0
```

**Comment by tacaswell** [comment]
Thank you for the detailed debugging work!

**Comment by tacaswell** [comment]
https://forum.hdfgroup.org/t/h5py-memory-leak-in-combination-with-pytorch-dataset-with-multiple-workers/9114 looks like it may be related.

**Comment by gvtulder** [comment]
I did some git bisect on the hdf5 repository, between [hdf5-1_12_0 (no leak) and hdf5-1_12_1 (leak)](https://github.com/HDFGroup/hdf5/compare/hdf5-1_12_0...hdf5-1_12_1).

Based on that, it looks like [this commit](https://github.com/HDFGroup/hdf5/commit/98a27f1fe4f0cb96ce7fdeee2e95079ae19db5cd) is where it goes wrong. Before, there's no leak, after, there is.

**Comment by tacaswell** [comment]
@gvtulder Are you taking care of reporting this upstream?

**Comment by gvtulder** [comment]
@tacaswell Sure, if you think that it's indeed an upstream bug.

Do you know if this isn't a case of h5py calling/not calling some function that shouldn't/should be called? I am not familiar with either the h5py or hdf5 code base, or the API. I can't say if this upstream change indeed introduced a leak, or whether it just made an existing bug in h5py visible.

In any case: I'd be happy to open an upstream issue if you think that is useful.

**Comment by tacaswell** [comment]
Thank you for reporting it upstream.

That is a fair point that it could be an error on the part of h5py that was exposed by this change.  However, in your example when you exit the context manager there should be only 1 Python object that knows about and hdf5 object left (the file and it should have been closed) which means nothing on the Python side should be holding things open.  When the function returns we no longer should have any Python side references to the underlying files.  At that point we should have also done all of the closing / deferencing we can do.  If we had a bug where we were not releasing everything then I would expect that to be relatively insensitive to the libhdf5 version.  It is definitely possible that we a leak on the h5py side, but given you were able to bisect it in libhdf5 and the commit message has the word "cache" in it, I would like upstream to say there is no obvious bug before we start digging in h5py!

**Comment by tacaswell** [comment]
Reading that diff, I am very suspicious of the changes to the reference counting on the VOL objects...

**Comment by recursix** [comment]
Hi, I have a very related bug with h5py 3.6 and 3.7. But for writing this time. If I don't use attrs, it doesn't leak. If I use attrs, it leaks. Is it expected to be fixed? Should I report as a separate issue?

**Comment by lbh666** [comment]
I met the same problem(memory leakage) when h5py installed with "pip".Then i uninstall h5py and reinstall it with "conda".The memory issue diappears.Hope this could be s solution for anyone who meets the same problem.

**Comment by NickLucche** [comment]
> I met the same problem(memory leakage) when h5py installed with "pip".Then i uninstall h5py and reinstall it with "conda".The memory issue diappears.Hope this could be s solution for anyone who meets the same problem.

Thanks same thing here, pip version 3.7.0. leaks for me while the conda binaries of the same version don't.

**Comment by tacaswell** [comment]
To be very clear, the issue is not directly related to how it is installed or to the version of h5py.  

We think that this is a bug in libhdf5 between 1.21.0 and 1.21.1 and the wheels on pypi and the conda packages are built with different versions of libhdf5.

I am going to close this because there is nothing that we can do on the h5py side to fix this.

**Comment by alvaro-budria** [comment]
I was facing the same issue with pip-installed `h5py==3.14.0` and a deep learning training pipeline.
I uninstalled h5py through pip, and reinstalled with mamba, and then the issue went away.

**Comment by neutrinoceros** [comment]
@alvaro-budria can you specify what OS + python version you saw this on ?

**Comment by tacaswell** [comment]
And most critically, the version of libhdf5.

**Comment by alvaro-budria** [comment]
Python version is `3.11.13`, OS is Red Hat `8.8` (Ootpa), libhdf5 is `1.14.6`.
